calyx/src/ir/guard.rs

```diff
@@ -52,13 +114,41 @@ impl Guard {
             Guard::Geq(_, _) => ">=".to_string(),
             Guard::Leq(_, _) => "<=".to_string(),
             Guard::Not(_) => "!".to_string(),
-            Guard::Port(_) => panic!("No operator string for Guard::Port"),
+            Guard::Port(_) | Guard::True => {
+                panic!("No operator string for Guard::Port")
+            }
         }
     }
 
-    ////////////// Convinience constructors ///////////////////
-    pub fn and(self, other: Guard) -> Self {
-        Guard::And(vec![self, other])
+    pub fn and_vec(&self, guards: &mut Vec<Guard>) -> Self {
+        flatten_and(guards);
+        let mut new;
+        if let Guard::And(inner) = &self {
+            new = inner.clone();
+            flatten_and(guards);
+            new.append(guards);
+        } else {
+            new = vec![self.clone()];
+            new.append(guards);
+        }
+        // filter out redundant guards
+        new.retain(|guard| {
+            if matches!(guard, Guard::True) {
+                return false;
+            }
+
+            if let Guard::Port(p) = guard {
+                if p.borrow().is_constant(1, 1) {
+                    return false;
+                }
+            }
```

Comment:
Move this to the top, and `return p.borrow().is_constant(1, 1)` followed by just returning `matches!(guard, Guard::True)`.